### DISCUSSION TITLE
How to install and use js packages

### DISCUSSION THREAD

**Discussion by shi-yan** [leptos-rs/leptos#912] [Q&A] [answered]
I'd love to use leptos for my front end, but I have some important dependences that are in js.

Some of them are complex, requiring me to provide callback functions that creates html elements 

Can I still use these js libraries?

**Comment by gbj** [comment, marked as answer]
I'd recommend taking a look at the [`wasm-bindgen` docs](https://rustwasm.github.io/docs/wasm-bindgen/examples/import-js.html) for how to import and call functions from JavaScript.

In general, the more a JS library wants to control your UI, the more difficult it will be to integrate it with any given UI framework, whether that's Leptos or a JS framework.

Without any more information that's about the level of detail it's possible to provide, but feel free to ask more specific questions as they arise.

**Reply by shi-yan** [reply]
Thank you for the answer! In particular, the dependence I need is a rich text editor. I'm currently using prose mirror, which requires me to provide call backs to supply things like context menu. 

rewriting it myself is too much work, I'm also not familiar with rich text editor implementation.

**Reply by jaads** [reply]
I have a similar problem, and I I'm not sure if wasm-bingen solves it. How would I integrate the NPM package in Laptop? Downloading it and storing it manually does not feel right. @shi-yan Did you solve it somehow with your rich text editor? I'd love to integrate `cytoscape.js`
